lib/fluent/plugin/buffered_output.rb

```diff
@@ -0,0 +1,522 @@
+        # c: constant factor, @retry_wait
+        # b: base factor, @retry_backoff_base
+        # k: times
+        # total retry time: c + c * b^1 + (...) + c*b^k = c*b^(k+1) - 1
+        config_param :retry_wait, :time, default: 1 # initial retry wait time & constant factor of exponencial backoff
+        config_param :retry_backoff_base, :float, default: 2 # base of exponencial backoff
+        config_param :retry_max_interval, :time, default: nil
+      end
+
+      config_section :secondary, param_name: :secondary_config, required: false, multi: false, final: true do
+        # to check there're <secondary> section at most once
+
+        config_section :buffer, required: false, multi: false do
+          # this section must not be specified in configuration files
+        end
+      end
+
+      def format(tag, time, record)
+        raise NotImplementedError, "BUG: buffered output plugins MUST implement this method"
+      end
+
+      ### Implement try_write if plugin does commit_flush under its own responsibility
+      # def try_write(chunk)
+
+      def write(chunk)
+        raise NotImplementedError, "BUG: buffered output plugins MUST implement this method"
+      end
+
+      def configure(conf)
+        super
+
+        c = @buffer_config
+
+        @buffer = Fluent::Plugin.new_buffer(c.type)
+        @buffer.configure(conf)
+
+        @flush_at_shutdown = c.flush_at_shutdown
+
+        @flush_interval = c.flush_interval
+        @flush_thread_interval = c.flush_thread_interval
+        @flush_burst_interval = c.flush_burst_interval
+        @flush_immediately = c.flush_immediately
+
+        @flush_threads = c.flush_threads
+
+        if @secondary_config
+          if @secondary_config.buffer
+            raise Fluent::ConfigError, "invalid <buffer> section configured in <secondary> section"
+          end
+          subconf = conf.elements.select {|e| e.name == 'secondary' }.first
+          # assume that secondary plugin is a same plugin to primary one if type isn't specified
+          subtype = subconf['@type'] || subconf['type'] || conf['@type'] || conf['type']
+          @secondary = Plugin.new_output(subtype)
+          @secondary.router = router
+          @secondary.configure(subconf)
+          @secondary.check_as_secondary(self)
+        end
+
+        @delayed_purge_enabled = false
+        @delayed_purge_timeout = nil
+      end
+
+      def enable_delayed_purge(timeout:)
+        unless self.respond_to?(:try_write)
+          raise Fluent::ConfigError, "BUG: try_write method required to enable delayed purge"
+        end
+        @delayed_purge_enabled = true
+        @delayed_purge_timeout = timeout
+      end
+
+      def check_as_secondary(primary)
+        if primary.class != self.class
+          log.warn "type of secondary output should be same as primary output", primary: primary.class.to_s, secondary: self.class.to_s
+        end
+      end
+
+      def start
+        @num_errors = 0
```

Comment:
Yes > intended
In my idea, main control factor for how long buffered plugins continue to retry, is time (minutes/hours), not retry times (but it will also be considered for compatibility).